device/transport/mqtt/devdoc/mqtt_twin_client_requirements.md

```diff
@@ -0,0 +1,95 @@
+# MqttTwinClient Requirements
+
+## Overview
+Object used to publish, subscribe and receive twin messages.
+
+## Responsibilities
+- subscribe to the necessary topics to achieve twin operations
+- format MQTT messages to send twin requests and reported properties updates
+- get the full twin using the appropriate topics and request/response mechanism
+- receive desired properties updates and signal them to the Mqtt client.
+
+## Non-responsibilities
+- connecting/disconnecting the Mqtt connection
+- any other feature that is not twin-related
+- retries and timeouts.
+
+## Public API
+```typescript
+class MqttTwinClient extends EventEmitter {
+  constructor(client: MqttBase);
+  getTwin(callback: (err?: Error, twin?: TwinProperties) => void): void;
+  updateTwinReportedProperties(patch: any, callback: (err?: Error) => void): void;
+  enableTwinDesiredPropertiesUpdates(callback: (err?: Error) => void): void;
+  disableTwinDesiredPropertiesUpdates(callback: (err?: Error) => void): void;
+}
+```
+
+### constructor(client: MqttBase);
+
+**SRS_NODE_DEVICE_MQTT_TWIN_CLIENT_18_001: [** The `MqttTwinClient` constructor shall accept a `client` object **]**
+
+**SRS_NODE_DEVICE_MQTT_TWIN_CLIENT_16_001: [** The `MqttTwinClient` constructor shall immediately subscribe to the `message` event of the `client` object. **]**
+
+### getTwin(callback: (err?: Error, twin?: TwinProperties) => void): void;
+
+The `getTwin` method retrieves the device twin by publishing a twin request on a topic and receiving the response on another one.
+
+**SRS_NODE_DEVICE_MQTT_TWIN_CLIENT_16_002: [** The `getTwin` method shall subscribe to the `$iothub/twin/res/#` topic if not already subscribed. **]**
+
+**SRS_NODE_DEVICE_MQTT_TWIN_CLIENT_16_009: [** If subscribing to the response topic fails, the callback shall be called with the translated version of the error obtained by using the `translateError` method of the `azure-iot-mqtt-base` package. **]**
+
+**SRS_NODE_DEVICE_MQTT_TWIN_CLIENT_16_003: [** The `getTwin` method shall publish the request message on the `$iothub/twin/get/?rid=<requestId>` topic using the `MqttBase.publish` method. **]**
+
+**SRS_NODE_DEVICE_MQTT_TWIN_CLIENT_16_004: [** The `getTwin` method shall publish the request message with QoS=0, DUP=0 and Retain=0. **]**
+
+**SRS_NODE_DEVICE_MQTT_TWIN_CLIENT_16_005: [** The `requestId` property in the topic querystring should be set to a random unique integer that will be used to identify the response later on. **]**
```

Comment:
>integer [](start = 130, length = 7)

could also be a guid.  that's what I should have done originally and that's what other SDKs do.

Reply:
I thought it had to be integers!! yes I like GUIDs better let me double check everything with the service 'cause i don't like generating random ints.

---
In reply to: [170696059](https://github.com/Azure/azure-iot-sdk-node/pull/236#discussion_r170696059) [](ancestors = 170696059)